src/Corax/Queries/SortingMatches/SortingMatch.Comparers.cs

```diff
@@ -0,0 +1,492 @@
+            readScores.Fill(Bm25Relevance.InitialScoreValue);
+
+            // We perform the scoring process. 
+            match._inner.Score(batchResults, readScores, 1f);
+
+            // If we need to do documents boosting then we need to modify the based on documents stored score. 
+            if (match._searcher.DocumentsAreBoosted)
+            {
+                // We get the boosting tree and go to check every document. 
+                BoostDocuments(match, batchResults, readScores);
+            }
+                
+            // Note! readScores & indexes are aliased and same as batchTermIds
+            var indexes = MemoryMarshal.Cast<long, int>(batchTermIds)[..(batchTermIds.Length)];
+            for (int i = 0; i < batchTermIds.Length; i++)
+            {
+                batchTerms[i] = new UnmanagedSpan(readScores[i]);
+                indexes[i] = i;
+            }
+
+            EntryComparerHelper.IndirectSort<EntryComparerByScore>(indexes, batchTerms, descending);
+
+            for (int i = 0; i < indexes.Length; i++)
+            {
+                match._results.Add(batchResults[indexes[i]]);
+            }
+        }
+
+        private static void BoostDocuments(SortingMatch<TInner> match, Span<long> batchResults, Span<float> readScores)
+        {
+            var tree = match._searcher.GetDocumentBoostTree();
+            if (tree is { NumberOfEntries: > 0 })
+            {
+                // We are going to read from the boosting tree all the boosting values and apply that to the scores array.
+                ref var scoresRef = ref MemoryMarshal.GetReference(readScores);
+                ref var matchesRef = ref MemoryMarshal.GetReference(batchResults);
+                for (int idx = 0; idx < batchResults.Length; idx++)
+                {
+                    var ptr = (float*)tree.ReadPtr(Unsafe.Add(ref matchesRef, idx), out var _);
+                    if (ptr == null)
+                        continue;
+
+                    ref var scoresIdx = ref Unsafe.Add(ref scoresRef, idx);
+                    scoresIdx *= *ptr;
+                }
+            }
+        }
+
+        public int Compare(UnmanagedSpan x, UnmanagedSpan y)
+        {
+            // Note, for scores, we go *descending* by default!
+            return y.Double.CompareTo(x.Double);
+        }
+    }
+
+    private struct CompactKeyComparer : IComparer<UnmanagedSpan>
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public int Compare(UnmanagedSpan xItem, UnmanagedSpan yItem)
+        {
+            if (yItem.Address == null)
+            {
+                return xItem.Address == null ? 0 : 1;
+            }
+
+            if (xItem.Address == null)
+                return -1;
+            var match = AdvMemory.Compare(xItem.Address + 1, yItem.Address + 1, Math.Min(xItem.Length - 1, yItem.Length - 1));
```

Comment:
Why `Math.Min(xItem.Length - 1, yItem.Length - 1)`? Aren't they different if they have different length?

Reply:
Because you would need to check that you are equal up to that point in time. If they are equal match is going to be 0 and then it is catched by the last few lines.

Reply:
We are searching _order_, not equality.